Upstream autocorrection enhancements support for WebKit

rr-codes opened the pull request:
#### 26c431674886169ee8198abcd8f3942c13dcb872
<pre>
Upstream autocorrection enhancements support for WebKit
<a href="https://bugs.webkit.org/show_bug.cgi?id=258410">https://bugs.webkit.org/show_bug.cgi?id=258410</a>
rdar://111141592

Reviewed by Aditya Keerthi.

Upstream support for autocorrection enhancements.

* Source/WTF/wtf/PlatformHave.h:
* Source/WTF/wtf/PlatformUse.h:
* Source/WebCore/editing/AlternativeTextController.cpp:
(WebCore::removeCorrectionIndicatorMarkers):
* Source/WebCore/editing/Editor.cpp:
(WebCore::shouldRemoveAutocorrectionIndicator):
(WebCore::didApplyAutocorrection):
(WebCore::respondToAppliedEditing):
(WebCore::adjustMarkerTypesToRemoveForWordsAffectedByEditing):
* Source/WebCore/platform/graphics/cocoa/GraphicsContextCocoa.mm:
(WebCore::drawRoundedRectForDocumentMarker):
(WebCore::GraphicsContextCG::drawDotsForDocumentMarker):
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::platformAutocorrectionReplacementMarkerColor const):
(WebCore::usePlatformColorForAutocorrectionReplacementMarker): Deleted.
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView applyAutocorrection:toString:shouldUnderline:withCompletionHandler:]):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::adjustCandidateAutocorrectionInFrame):

Canonical link: <a href="https://commits.webkit.org/265489@main">https://commits.webkit.org/265489@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7dc3ff36c708f1b03feeb8d3b16ddcf4c3edb4d2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/10872 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/11088 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/11414 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/12520 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/10408 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/10888 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/13462 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/11067 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/13322 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/11031 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/11937 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/9173 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/12924 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/9230 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/9831 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/17059 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/9208 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/10301 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9983 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/13212 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/10314 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/10426 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/8511 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/10982 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/9591 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/2996 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13864 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/11290 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1236 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/10287 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/2760 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/editing/Editor.cpp:
```diff
@@ -1125,30 +1125,72 @@ static void notifyTextFromControls(Element* startRoot, Element* endRoot)
 {
     alternativeTextController.respondToAppliedEditing(&command);
+#if HAVE(AUTOCORRECTION_ENHANCEMENTS) && PLATFORM(IOS_FAMILY)
+    // FIXME: This should be moved into AlternativeTextController when upstreamed.
```
This `FIXME` should be updated, since it's not happening in this PR.

rr-codes replied:
should I just do what the comment says and move it into AlternativeTextController instead?

pxlcoder replied:
Sounds good, should be pretty simple to move into `AlternativeTextController::respondToAppliedEditing`.

pxlcoder reviewed (approved):
I think there are some opportunities for refactoring/renaming (for example, `isCandidate` -> `shouldUnderline`), but this is a good upstream.

pxlcoder commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -10915,6 +10915,11 @@ - (BOOL)supportsTextReplacementForWebView
     return self.webView._editable;
 }
 
+- (void)applyAutocorrection:(NSString *)correction toString:(NSString *)input shouldUnderline:(BOOL)shouldUnderline withCompletionHandler:(void (^)(UIWKAutocorrectionRects *rectsForCorrection))completionHandler
```
Can you move this closer to the existing `applyAutocorrection:` methods?

rr-codes replied:
yes! I was looking for a better place to put it

rr-codes committed:
Upstream autocorrection enhancements support for WebKit
https://bugs.webkit.org/show_bug.cgi?id=258410
rdar://111141592

Reviewed by Aditya Keerthi.

Upstream support for autocorrection enhancements.

* Source/WTF/wtf/PlatformHave.h:
* Source/WTF/wtf/PlatformUse.h:
* Source/WebCore/editing/AlternativeTextController.cpp:
(WebCore::removeCorrectionIndicatorMarkers):
* Source/WebCore/editing/Editor.cpp:
(WebCore::shouldRemoveAutocorrectionIndicator):
(WebCore::didApplyAutocorrection):
(WebCore::respondToAppliedEditing):
(WebCore::adjustMarkerTypesToRemoveForWordsAffectedByEditing):
* Source/WebCore/platform/graphics/cocoa/GraphicsContextCocoa.mm:
(WebCore::drawRoundedRectForDocumentMarker):
(WebCore::GraphicsContextCG::drawDotsForDocumentMarker):
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::platformAutocorrectionReplacementMarkerColor const):
(WebCore::usePlatformColorForAutocorrectionReplacementMarker): Deleted.
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView applyAutocorrection:toString:shouldUnderline:withCompletionHandler:]):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::adjustCandidateAutocorrectionInFrame):

Canonical link: https://commits.webkit.org/265489@main
